shared-bindings/keypad/EventQueue.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "shared-bindings/keypad/Event.h"
+#include "shared-bindings/keypad/EventQueue.h"
+
+//| class EventQueue:
+//|     """A queue of `Event` objects, filled by a `keypad` scanner such as `Keys` or `KeyMatrix`.
+//|
+//|     You cannot create an instance of `EventQueue` directly. Each scanner creates an
+//|     instance when it is created.
+//|     """
+//|     ...
+
+//|     def next(self) -> Optional[Event]:
+//|         """Return the next key transition event. Return ``None`` if no events are pending.
+//|
+//|         Note that the queue size is limited; see ``max_events`` in the constructor of
+//|         a scanner such as `Keys` or `KeyMatrix`.
+//|         If a new event arrives when the queue is full, the oldest event is discarded.
```

Comment:
Do we want to inject an overflow event to mark where the event stream is disjoint? You'd probably want to treat it as if all keys were released.

Reply:
We could have an overflow event, or just a resettable flag. I'm not sure we want to release all the keys because we lose a lot of state that way. For instance, if someone is pounding a controller key with a shift key held down, it's OK to lose some of those clicks, but not the shift key state.

Reply:
But if they release the shift at that moment, you will have it wrongly held forever. Resetting to a known good state is the only way to handle this.

Reply:
I defer to your experience on the clearing.

I added an `overflowed` property to `Event`, and I will clear the `EventQueue` and post that event if it overflows.

Reply:
Do you want to repopulate the eventqueue with current presses on overflow? You still know the current state since you maintain it separately.

Reply:
But what if there are more pressed keys than the queue size?